File: apps/android/app/src/main/kotlin/com/wordocious/app/data/ConsentGate.kt
package com.wordocious.app.data

import android.content.Context
import android.telephony.TelephonyManager
import java.util.Locale

/**
 * Region gate for the ads SDK — docs/LEVELPLAY_SETUP.md "Consent".
 *
 * Unity LevelPlay ships no consent-management platform of its own (AppLovin
 * MAX did; Google's UMP is unusable because its forms live in the dead AdMob
 * console). Until a CMP is integrated the app must not initialize the ads SDK
 * for users in a region that requires prior consent: those users simply stay
 * ad-free. The decision is made from the device region and nothing else.
 *
 * Phase 2 (a separate task): integrate Usercentrics (Unity's partner CMP),
 * paste its settings id here, and lift the gate once the CMP has run.
 */
object ConsentGate {
    /**
     * Regions whose users see no ads until a CMP exists: the 27 EU / EEA
     * member states, the three EFTA states in the EEA (Norway, Iceland,
     * Liechtenstein), the United Kingdom and Switzerland. ISO 3166-1 alpha-2,
     * upper case; "UK" is not an ISO code but is included defensively in case
     * a vendor locale uses it.
     */
    private val CONSENT_REGIONS: Set<String> = setOf(
        // EU 27
        "AT", "BE", "BG", "HR", "CY", "CZ", "DK", "EE", "FI", "FR", "DE", "GR", "HU",
        "IE", "IT", "LV", "LT", "LU", "MT", "NL", "PL", "PT", "RO", "SK", "SI", "ES", "SE",
        // EEA / EFTA
        "NO", "IS", "LI",
        // UK GDPR, Swiss FADP
        "GB", "UK", "CH",
    )

    /**
     * Best-effort device country: the default locale's region first (what the
     * user set), the mobile network's country as the fallback when the locale
     * carries no region (e.g. a bare "en"). Empty when neither is known.
     */
    fun deviceCountry(context: Context): String {
        val fromLocale = Locale.getDefault().country.orEmpty()
        if (fromLocale.isNotEmpty()) return fromLocale.uppercase(Locale.ROOT)
        val fromNetwork = runCatching {
            context.getSystemService(TelephonyManager::class.java)?.networkCountryIso
        }.getOrNull().orEmpty()
        return fromNetwork.uppercase(Locale.ROOT)
    }

    /** True when the ads SDK must NOT be initialized on this device. */
    fun blocksAds(context: Context): Boolean = deviceCountry(context) in CONSENT_REGIONS
}

File: apps/ios/Wordocious/Sources/ConsentGate.swift
import Foundation

/// Decides whether the ads SDK may initialize at all, from the device region.
///
/// LevelPlay ships no consent-management platform (AppLovin MAX did; Google's
/// UMP lives inside the dead AdMob console). Until a CMP is integrated, users
/// in GDPR-style consent regions simply never see the SDK start — the region
/// stays ad-free rather than serving without a lawful consent record.
///
/// Phase 2 (docs/LEVELPLAY_SETUP.md): a Usercentrics App CMP account lifts the
/// gate for these regions; that is a separate task, and this type is where its
/// settings id will land.
enum ConsentGate {
    enum Decision: Equatable {
        /// Outside every consent region — the SDK may start.
        case allowed
        /// Inside the EEA / UK / Switzerland / EFTA — do not initialize ads.
        case blocked
    }

    /// EU-27 + the EEA EFTA states (Norway, Iceland, Liechtenstein) + UK +
    /// Switzerland. ISO 3166-1 alpha-2, as `Locale.Region.identifier` returns.
    static let consentRegions: Set<String> = [
        // EU-27
        "AT", "BE", "BG", "HR", "CY", "CZ", "DK", "EE", "FI", "FR", "DE", "GR",
        "HU", "IE", "IT", "LV", "LT", "LU", "MT", "NL", "PL", "PT", "RO", "SK",
        "SI", "ES", "SE",
        // EEA (EFTA)
        "NO", "IS", "LI",
        // UK + Switzerland
        "GB", "CH",
    ]

    /// Decision for the current device region. An unknown region blocks — the
    /// failure mode of "no ads" is the safe one.
    static var decision: Decision { decision(forRegion: Locale.current.region?.identifier) }

    static func decision(forRegion code: String?) -> Decision {
        guard let code = code?.uppercased(), !code.isEmpty else { return .blocked }
        return consentRegions.contains(code) ? .blocked : .allowed
    }
}
